libs/vkd3d/swapchain.c

```diff
@@ -0,0 +1,2216 @@
+        return S_OK;
+
+    memset(&image_info, 0, sizeof(image_info));
+    image_info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
+    image_info.flags = VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT;
+    image_info.imageType = VK_IMAGE_TYPE_2D;
+    image_info.format = vk_format;
+    image_info.extent.width = swapchain->desc.Width;
+    image_info.extent.height = swapchain->desc.Height;
+    image_info.extent.depth = 1;
+    image_info.mipLevels = 1;
+    image_info.arrayLayers = 1;
+    image_info.samples = VK_SAMPLE_COUNT_1_BIT;
+    image_info.tiling = VK_IMAGE_TILING_OPTIMAL;
+    image_info.usage = VK_IMAGE_USAGE_TRANSFER_SRC_BIT
+            | VK_IMAGE_USAGE_TRANSFER_DST_BIT
+            | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
+    image_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+    image_info.queueFamilyIndexCount = 0;
+    image_info.pQueueFamilyIndices = NULL;
+    image_info.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+
+    for (i = 0; i < swapchain->desc.BufferCount; ++i)
+    {
+        VkMemoryDedicatedRequirements dedicated_requirements;
+        VkImageMemoryRequirementsInfo2 image_requirements_info;
+        VkMemoryDedicatedAllocateInfo dedicated_allocate_info;
+        VkMemoryAllocateInfo allocate_info;
+        VkMemoryRequirements2 requirements;
+        uint32_t memory_type_index;
+
+        assert(swapchain->vk_images[i] == VK_NULL_HANDLE);
+        if ((vr = vk_procs->vkCreateImage(vk_device, &image_info, NULL, &swapchain->vk_images[i])) < 0)
+        {
+            WARN("Failed to create Vulkan image, vr %d.\n", vr);
+            swapchain->vk_images[i] = VK_NULL_HANDLE;
+            return hresult_from_vk_result(vr);
+        }
+
+        dedicated_requirements.sType = VK_STRUCTURE_TYPE_MEMORY_DEDICATED_REQUIREMENTS;
+        dedicated_requirements.pNext = NULL;
+
+        requirements.sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2;
+        requirements.pNext = &dedicated_requirements;
+
+        image_requirements_info.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_REQUIREMENTS_INFO_2;
+        image_requirements_info.pNext = NULL;
+        image_requirements_info.image = swapchain->vk_images[i];
+
+        vk_procs->vkGetImageMemoryRequirements2(vk_device,
+            &image_requirements_info, &requirements);
+
+        dedicated_allocate_info.sType  = VK_STRUCTURE_TYPE_MEMORY_DEDICATED_ALLOCATE_INFO;
+        dedicated_allocate_info.pNext  = NULL;
+        dedicated_allocate_info.image  = swapchain->vk_images[i];
+        dedicated_allocate_info.buffer = VK_NULL_HANDLE;
+
+        if (FAILED(hr = vk_select_memory_type(vk_procs, vk_physical_device,
+            requirements.memoryRequirements.memoryTypeBits,
+            VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT, &memory_type_index)))
+            return hr;
+
+        allocate_info.sType           = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+        allocate_info.pNext           = dedicated_requirements.prefersDedicatedAllocation
```

Comment:
We're going to allocate once per image anyways, why bother to check for dedicated prefer? I'd suggest just always making a dedicated allocation here.

Reply:
Changed to use vkd3d resource creation.